crates/valence_core/src/text/color.rs

```diff
@@ -0,0 +1,368 @@
+    Rgb(RgbColor),
+    /// One of the 16 normal Minecraft colors
+    Normal(NormalColor),
+}
+
+/// RGB Color
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct RgbColor {
+    /// Red channel
+    pub r: u8,
+    /// Green channel
+    pub g: u8,
+    /// Blue channel
+    pub b: u8,
+}
+
+/// Normal Minecraft color
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum NormalColor {
+    /// Hex digit: `0`, name: `black`
+    Black = 0,
+    /// Hex digit: `1`, name: `dark_blue`
+    DarkBlue,
+    /// Hex digit: `2`, name: `dark_green`
+    DarkGreen,
+    /// Hex digit: `3`, name: `dark_aqua`
+    DarkAqua,
+    /// Hex digit: `4`, name: `dark_red`
+    DarkRed,
+    /// Hex digit: `5`, name: `dark_purple`
+    DarkPurple,
+    /// Hex digit: `6`, name: `gold`
+    Gold,
+    /// Hex digit: `7`, name: `gray`
+    Gray,
+    /// Hex digit: `8`, name: `dark_gray`
+    DarkGray,
+    /// Hex digit: `9`, name: `blue`
+    Blue,
+    /// Hex digit: `a`, name: `green`
+    Green,
+    /// Hex digit: `b`, name: `aqua`
+    Aqua,
+    /// Hex digit: `c`, name: `red`
+    Red,
+    /// Hex digit: `d`, name: `light_purple`
+    LightPurple,
+    /// Hex digit: `e`, name: `yellow`
+    Yellow,
+    /// Hex digit: `f`, name: `white`
+    White,
+}
+
+/// Color parsing error
+#[derive(Debug, Error, PartialEq, PartialOrd, Clone, Copy, Hash, Eq, Ord)]
+#[error("invalid color name or hex code")]
+pub struct ColorError;
+
+impl Color {
+    pub const RESET: Self = Self::Reset;
+    pub const AQUA: Self = Self::Normal(NormalColor::Aqua);
+    pub const BLACK: Self = Self::Normal(NormalColor::Black);
+    pub const BLUE: Self = Self::Normal(NormalColor::Blue);
+    pub const DARK_AQUA: Self = Self::Normal(NormalColor::DarkAqua);
+    pub const DARK_BLUE: Self = Self::Normal(NormalColor::DarkBlue);
+    pub const DARK_GRAY: Self = Self::Normal(NormalColor::DarkGray);
+    pub const DARK_GREEN: Self = Self::Normal(NormalColor::DarkGreen);
+    pub const DARK_PURPLE: Self = Self::Normal(NormalColor::DarkPurple);
+    pub const DARK_RED: Self = Self::Normal(NormalColor::DarkRed);
+    pub const GOLD: Self = Self::Normal(NormalColor::Gold);
+    pub const GRAY: Self = Self::Normal(NormalColor::Gray);
+    pub const GREEN: Self = Self::Normal(NormalColor::Green);
+    pub const LIGHT_PURPLE: Self = Self::Normal(NormalColor::LightPurple);
+    pub const RED: Self = Self::Normal(NormalColor::Red);
+    pub const WHITE: Self = Self::Normal(NormalColor::White);
+    pub const YELLOW: Self = Self::Normal(NormalColor::Yellow);
+
+    /// Constructs a new RGB color
+    pub const fn rgb(r: u8, g: u8, b: u8) -> Self {
+        Self::Rgb(RgbColor::new(r, g, b))
+    }
+}
+
+impl RgbColor {
+    /// Constructs a new color from red, green, and blue components.
+    pub const fn new(r: u8, g: u8, b: u8) -> Self {
+        Self { r, g, b }
+    }
+}
+
+impl NormalColor {
+    pub const fn as_hex_digit(self) -> char {
```

Comment:
Since `NormalColor` is `Copy` maybe the best solution here would be simple `fn hex_digit(self)` and `fn name(self)` without either `as` or `to` to avoid confusion?

Reply:
I would associate `fn hex_digit(self)` with an accessor while `fn to_hex_digit(self)` with a function that does some work to convert to another type.

https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv